### ISSUE TITLE
Documentation for Dart SIMD missing

### ISSUE THREAD

**Issue by maks** [dart-lang/site-www#3679]
### What information is missing?

There is not really any information at all regarding Dart supporting SIMD, in fact you would need to be fairly determined in searching dart.dev to even discover that such support exists.

Yet there used to be what I would consider a pretty good documentation page for it: https://web.archive.org/web/20190402213240/https://www.dartlang.org/articles/server/simd

and reading through it, I am unclear as to why it was considered obsolete and hence removed from the website? Apart from its use of the `new` keyword, is any of the code or the prose text actually inaccurate for the current versions of Dart lang?

### How would you like us to fix this problem?

Please re-instate the previously removed page documenting SIMD usage in Dart.

**Comment by maks** [comment]
I would also be happy to contribute a PR updating the previous documentation page if someone could give some guidance on what specifically was lacking in it that caused it to be classified as "obsolete".

**Comment by parlough** [comment]
Thanks for sharing your interest in this page @maks. It would be nice to further SIMD documentation, perhaps by starting with this and making some updates. I'll ping @kwalrath and @mit-mit to see their thoughts on reviving an updated version of this page as it was originally removed in https://github.com/dart-lang/site-www/pull/1572

**Comment by kwalrath** [comment]
@mit-mit what would you like us to do about SIMD? It seems like we could update the article, put it under https://dart.dev/guides/libraries, and add mentions of SIMD to https://dart.dev/guides/language/numbers & maybe the language/library tour (at least).

What do you think? Who knows the most about SIMD now?

**Comment by kwalrath** [comment]
@maks I think the code examples would need to pass analysis using the current set of recommended lints. We'd ensure this by adding all code in the article to [/examples](https://github.com/dart-lang/site-www/tree/master/examples), as documented in https://github.com/dart-lang/site-shared/blob/master/doc/code-excerpts.md & https://github.com/dart-lang/site-shared/blob/master/doc/examples.md.

**Comment by maks** [comment]
@kwalrath thanks for looking into this so quickly! I'll have a read of those docs to see how it's done - or should I wait for @mit-mit to say on if the SIMD docs can come back before I start working on it? 
My 2c about why I raised this is that I think the fact that Dart supports SIMD is really a hidden gem of Dart that very few devs I spoken with in the community know about and would be a great selling point for Dart usage for some applications.

**Comment by kwalrath** [comment]
I suspect we'd be very happy to accept any help you can offer... but I can't guarantee that without talking to @mit-mit. Or maybe to @kevmoo.

So if you feel like looking at the text or playing with the code, that'd be awesome, but I can't make any promises about using the results... yet.

**Comment by kevmoo** [comment]
Most of our SIMD support IIRC is just methods in https://api.dart.dev/stable/2.14.4/dart-typed_data/dart-typed_data-library.html

Things like https://api.dart.dev/stable/2.14.4/dart-typed_data/Int32x4/operator_plus.html

Where you can operate on "chunks" of numeric values together.

**Comment by maks** [comment]
@kevmoo thats true, but the dartdoc documentation is somewhat terse. Really I just found that the previous documentation page did a really nice job of clearly explaining SIMD and how to make use of SIMD in the context of Dart.  

Having something like this article that brings it all together with some diagrams and small code examples (at least for me) I find is super helpful if being able to dive in and start using it quickly. For example the explanation and diagrams in the [branching section](https://web.archive.org/web/20190402213240/https://www.dartlang.org/articles/server/simd#branching) I found helpful as a newcomer to using SIMD (not just SIMD in Dart).

**Comment by kwalrath** [comment]
> Most of our SIMD support IIRC is just methods in https://api.dart.dev/stable/2.14.4/dart-typed_data/dart-typed_data-library.html
> 
> Things like https://api.dart.dev/stable/2.14.4/dart-typed_data/Int32x4/operator_plus.html
> 
> Where you can operate on "chunks" of numeric values together.

There was a [really good article on SIMD](https://web.archive.org/web/20190402213240/https://www.dartlang.org/articles/server/simd) that I suspect we jettisoned because its code was untested and rotting. The figures alone are very helpful!

**Comment by atsansone** [comment]
@mit-mit , @kevmoo : Is this something you would like to see documented in the current public docs? If so, it would need updates for Dart 2 if not Dart 3. Thoughts?

**Comment by kevmoo** [comment]
@atsansone – I don't there have been any changes w/ Dart 3.

**Comment by maks** [comment]
My main reason for opening this issue originally was just that I found https://web.archive.org/web/20190402213240/https://www.dartlang.org/articles/server/simd really helpful and useful and thought others would too. Would the Dart team be open to republishing on dart.dev again if I submitted a PR which updated the code examples in it to modern Dart usage (eg removing `new`'s)?

**Comment by mit-mit** [comment]
@mraleph can you take a quick look at https://web.archive.org/web/20190402213240/https://www.dartlang.org/articles/server/simd and determine if that content is still relevant for the current-day VM?

**Comment by mraleph** [comment]
The content is okay if it is updated. It's not just about updating the code - e.g. it should reflect that `SIMD.js` is abandoned, etc. 

That being said, `dart:typed_data` API has some issues and I am not sure we want to evangelise it. If anything we should explore the way to pull it out of `dart:typed_data` and instead provide a number of platform specific intrinsic libraries. There are two core issues with the current API:

- It was designed in an Intel centric manner - however ARM and Intel instruction sets have different capabilities where it comes to SIMD e.g. classical example would be `signMask` which is a single instruction on Intel but requires multiple instructions on Arm. Consquently some SIMD operations have really bad implementations on ARM platforms.  
- It does not have and will never have a performant JS implementation, despite residing in a platform independent library.

**Comment by maks** [comment]
Thanks for looking into it @mraleph 
Good to know about this limitations of the API, I'm still a beginner in SIMD (hence why I liked having the old article available) so I didn't realise about it being Intel centric.

I understand not wanting to promote something that you would like to replace, but I guess the current API is going to be here for quite sometime or are there plans for this to change this soon? sorry I couldn't seem to find if there is a roadmap for DartVM features vs the one that exists for upcoming language features/changes.

On the web front, would the upcoming support in Dart for WASM also help with not needing to worry as much about the JS implementation? Funny thing is that the [V8 doc on WASM SIMD ](https://v8.dev/features/simd) actually references as being influenced by both SIMD.js and the Dart SIMD 🙂 .

**Comment by kevmoo** [comment]
If my memory serves, we decided to drop the SIMD documentation as it was poorly maintained and outdated.

**Comment by isoos** [comment]
I think any non-perfect documentation is better than no documentation (or one that will change in the future), otherwise googling for "dart simd" will yield a parody-site as the first result (like as of now).
